### PR TITLE
Ensure that the reloading of a config file only happenes once

### PR DESCRIPTION
**Pull request by jrouzierinverse**
## DESCRIPTION

Ensure that configuration files are only reloaded once to avoid reparsing
## Impacts

This impact all services that use configuration
bin/pfcmd configreload
bin/pfcmd configreload hard
## NEWS file entries

Bug Fixes
+++++++++
Fixed issue with cache reloading increasing memory usage while reloading the cache
## Delete branch after merge

NO

### CODE REVIEW

**Comment by julsemaan** [comment]
Minor detail but can you change the title, it's DESCRIPTION

This way it's easier to refer to it or search it when it's closed.

**Comment by jrouzierinverse** [comment]
Thanks

**Comment by julsemaan** [comment]
@fdurand we'll need the packages built before we can merge this.

I'll test it in the meantime

**Comment by jrouzierinverse** [comment]
The updated CHI is built already
File::FcntlLock just needs to be built 
You can get it from epel in the mean time perl-File-FcntlLock

File: addons/dev-helpers/config-reload-loop.pl [hunk 1]
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/perl
+
+=head1 NAME
+
+loop.pl - a script to help test reloading config
```
Comment (cgx) [inline addons/dev-helpers/config-reload-loop.pl]: It doesn't match the exact name.

File: addons/dev-helpers/config-reload-loop.pl [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+=cut
+
+=head1 DESCRIPTION
+
+loop
```
Comment (cgx) [inline addons/dev-helpers/config-reload-loop.pl]: Can you be more precise?

File: lib/pf/FileLocker.pm [hunk 3]
```diff
@@ -0,0 +1,229 @@
+package pf::FileLocker;
+=head1 NAME
+
+pf::FileLocker add documentation
```
Comment (cgx) [inline lib/pf/FileLocker.pm]: Fix NAME and DESCRIPTION sections of the POD.

File: lib/pf/config/cached.pm [hunk 4]
```diff
@@ -835,6 +916,23 @@ sub cacheForData {
...
+sub RefreshConfigs {
+    my $logger = get_logger();
+    $logger->trace("Started Reloading all configs");
+    foreach my $config (@LOADED_CONFIGS{@LOADED_CONFIGS_FILE}) {
+        next unless $config;
+        next unless $config->LockFileHasChanged;
+        $config->ReloadConfig();
+    }
+    $logger->trace("Finished Reloading all configs");
+}
```
Comment (cgx) [inline lib/pf/config/cached.pm:940]: Your log entries for this sub are the same as for `ReloadConfigs`. You should reword them.

**Comment by julsemaan** [comment]
Tested with https://gist.github.com/julsemaan/302783b46c098e8fcb34 
- 2 processes updating an entry in the switches.conf file
- tested with 2915 switches
- Both are able to interact with the config store and load is not escalating too much.

Admin interface is slow but everything is still working.

**Comment by julsemaan** [comment]
the chi version on the repo is 0.58-2 and not 0.59

You'll need to adjust the spec file

**Comment by jrouzierinverse** [comment]
Check the packetfence-devel repo

**Comment by julsemaan** [comment]
I did

**Comment by julsemaan** [comment]
Confirmed : combining a switch modification with radius stress testing makes the load go high and almost no radius request are answered (using 2000 switches)

**Comment by julsemaan** [comment]
What is everybody's opinion on this PR ?

Should we merge it or leave the code as is for now ?

**Comment by jrouzierinverse** [comment]
Please Review or Merge

**Comment by julsemaan** [comment]
No relevent anymore IMO. Close ?

**Comment by louismunro** [comment]
Is the performance issue still there? 
If it's still there let's close this and rework it.

**Comment by julsemaan** [comment]
The performance issue is still there but we're rewriting the config access to I'm not sure we should be adding additionnal layers to the 'old' config

**Comment by louismunro** [comment]
Agreed.
Let's close until we know better where we are going with this.
We could always reopen if needed.

**Pull request closed without merging**